src/ingest/src/main/scala/osmesa/ingest/ProcessOSM.scala

```diff
@@ -0,0 +1,399 @@
+
+      val geom = coords match {
+        // drop ways with invalid coordinates
+        // check for nulls (or add a coalesce to the select after join); since posexplode_outer is used, there will always be
+        // a row but it may contain null (not NaN) values
+        case _ if coords.exists { case (x, y) => x == null || x.equals(Double.NaN) || y == null || y.equals(Double.NaN) } => None
+        // drop ways that don't contain valid geometries
+        case _ if coords.isEmpty => None
+        case _ if coords.length == 1 =>
+          Some(Point(coords.head._1, coords.head._2))
+        case _ =>
+          val line = Line(coords)
+
+          Some(isArea match {
+            case true if line.isClosed && coords.length >= 4 => Polygon(line)
+            case _ => line
+          })
+      }
+
+      geom match {
+        // drop invalid geometries
+        case Some(g) if g.isValid => g.toWKB(4326)
+        case _ => null
+      }
+    })
+
+    // useful for debugging
+    val isValid = udf((geom: Array[Byte]) => {
+      geom match {
+        case null => true
+        case _ => geom.readWKB.isValid
+      }
+    })
+
+    // useful for debugging; some geometries that are valid as 4326 are not as 3857
+    val reproject = udf((geom: Array[Byte]) => {
+      geom match {
+        case null => null
+        case _ => geom.readWKB.reproject(LatLng, WebMercator).toWKB(3857)
+      }
+    })
+
+    // useful for debugging
+    val asWKT = udf((geom: Array[Byte]) => {
+      geom match {
+        case null => ""
+        case _ => geom.readWKB.toWKT
+      }
+    })
+
+    // Create WKB geometries (LineStrings and Polygons)
+    val wayGeoms = taggedWays
+      .withColumn("geom", asWKB('coords, isArea('tags)))
+      .select('changeset, 'id, 'version, 'tags, 'geom, 'updated, 'visible)
+
+    // Assign `minorVersion` and rewrite `validUntil` to match
+    @transient val idAndVersionByUpdated = Window.partitionBy('id, 'version).orderBy('updated)
+    @transient val idByUpdated = Window.partitionBy('id).orderBy('updated)
+
+    wayGeoms
+      .withColumn("validUntil", lead('updated, 1) over idByUpdated)
+      .withColumn("minorVersion", (row_number() over idAndVersionByUpdated) - 1)
+      .select('changeset, 'id, 'version, 'tags, 'geom, 'updated, 'validUntil, 'visible, 'minorVersion)
+  }
+
+  def geometriesByRegion(nodeGeoms: Dataset[Row], wayGeoms: Dataset[Row], useS3: Boolean): DataFrame = {
+    import nodeGeoms.sparkSession.implicits._
+
+    // Geocode geometries by country
+    val geoms = nodeGeoms.union(wayGeoms)
+
+    object Resource {
+      def apply(name: String): String = {
+        val stream: InputStream = getClass.getResourceAsStream(s"/$name")
+        try { scala.io.Source.fromInputStream( stream ).getLines.mkString(" ") } finally { stream.close() }
+      }
+    }
+
+    case class CountryId(code: String)
+
+    object MyJsonProtocol extends DefaultJsonProtocol {
```

Comment:
There's another way to do this in https://github.com/azavea/osmesa/blob/e84e37be91535607cb9b0d6758375cdb7263abe6/src/analytics/src/main/scala/osmesa/analytics/Countries.scala#L13-L29

It's done this way here because the `Countries.scala` approach didn't work w/ Zeppelin, as fragments are effectively `:paste`d into the interpreter.

Reply:
I'm supposing, based on your commenting, that you'd prefer to use the above?

Reply:
No preference; I was cargo culting at this point to get it working. Mainly flagging that there was no good reason for it being this way.